common/payload/payload.go

```diff
@@ -94,11 +95,28 @@ func MergeMapOfPayload(
 	}
 	res := util.CloneMapNonNil(dst)
 	for k, v := range src {
-		if proto.Equal(v, nilPayload) || proto.Equal(v, emptySlicePayload) {
+		if isEqual(v, nilPayload) || isEqual(v, emptySlicePayload) {
 			delete(res, k)
 		} else {
 			res[k] = v
 		}
 	}
 	return res
 }
+
+// isEqual returns true if both have the same encoding and data.
+// It does not take additional metadata into consideration.
+func isEqual(a, b *commonpb.Payload) bool {
+	return getEncoding(a) == getEncoding(b) && bytes.Equal(a.GetData(), b.GetData())
+}
+
+func getEncoding(payload *commonpb.Payload) string {
+	metadata := payload.GetMetadata()
+	if metadata == nil {
+		return ""
+	}
+	if e, ok := metadata[converter.MetadataEncoding]; ok {
+		return string(e)
+	}
+	return ""
```

Comment:
```suggestion
	return payload.GetMetadata()[converter.MetadataEncoding]
```

No need to complicate it (probably don't need a new method)

Reply:
Oh, I thought it would panic if the map is nil, but you're right, it returns an empty string.